examples/huggingface/pytorch/text-generation/inference/run_generation_with_deepspeed.py

```diff
@@ -0,0 +1,578 @@
+# dtype = torch.bfloat16 if model_name in ["bigscience/bloom", "bigscience/bigscience-small-testing"] else torch.float16
+
+
+# use one of these args to `init_inference`
+# 1. injection_policy is the slower version, but it's plain pytorch so it'll always work
+# 2. replace_with_kernel_inject is the faster one (fast fused kernels)
+kernel_inject = args.ki
+
+if args.benchmark:
+    get_accelerator().empty_cache()
+    gc.collect()
+    deepspeed.runtime.utils.see_memory_usage("pre-from-pretrained", force=True)
+
+is_meta_support = not model_type in ["falcon"]
+
+# Construct model with fake meta tensors, later will be replaced during ds-inference ckpt load
+with deepspeed.OnDevice(dtype=load_dtype, device="meta", enabled=is_meta_support):
+    # Even inside the meta device context, from_pretrained still loads the
+    # model to cpu instead of meta device. Use from_config instead to solve the issue for big models.
+    # We add the instance type check here since some of the models haven't yet supported from_config.
+    if model_class[0] == AutoModelForCausalLM and is_meta_support:
+        model = model_class[0].from_config(config, torch_dtype=load_dtype, trust_remote_code=True)
+    else:
+        model = model_class[0].from_pretrained(model_name, config=config, low_cpu_mem_usage=True, torch_dtype=load_dtype, trust_remote_code=True)
+
+if args.benchmark:
+    deepspeed.runtime.utils.see_memory_usage("post-from-pretrained", force=True)
+
+model = model.eval()
+model = model.to(memory_format=torch.channels_last)
+
+if args.benchmark:
+    get_accelerator().empty_cache()
+    gc.collect()
+    deepspeed.runtime.utils.see_memory_usage("post-init-ds-zero-init", force=True)
+
+### Deepspeed-Inference Loading
+
+checkpoints_json = "checkpoints.json"
+
+
+def write_checkpoints_json():
+    checkpoint_files = get_checkpoint_files(model_name)
+    if local_rank == 0:
+        data = {"type": "BLOOM", "checkpoints": checkpoint_files, "version": 1.0}
+        json.dump(data, open(checkpoints_json, "w"))
+
+
+if args.benchmark:
+    get_accelerator().empty_cache()
+    gc.collect()
+    deepspeed.runtime.utils.see_memory_usage("pre-ds-inference-init", force=True)
+
+if kernel_inject:
+    kwargs = dict(replace_with_kernel_inject=True)
+else:
+    kwargs = dict(replace_with_kernel_inject=False)
+
+repo_root = get_repo_root(model_name)
+if tp_presharded_mode and kernel_inject:
+    # tp presharded repos come with their own checkpoints config file
+    checkpoints_json = os.path.join(repo_root, "ds_inference_config.json")
+else:
+    # for normal bloom repo we need to write the checkpoints config file
+    write_checkpoints_json()
+    dist.barrier()
+
+model = deepspeed.init_inference(
+    model,
+    mp_size=world_size,
+    base_dir=repo_root,
+    dtype=infer_dtype,
+    checkpoint=checkpoints_json if is_meta_support else None,
+    **kwargs,
+)
+
+if args.benchmark:
+    get_accelerator().empty_cache()
```

Comment:
these 3 lines are frequently used, suggest to be a util func for caller use.

Reply:
done